config/reminder-config.yaml.example

```diff
@@ -0,0 +1,40 @@
+#
+# Copyright 2024 Stacklok, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+recurrence:
+  interval: "1h"
+  batch_size: 100
+  max_per_project: 10
+  min_project_fetch_limit: 10
+  min_elapsed: "1h"
+
+database:
+  dbhost: "postgres"
+  dbport: 5432
+  dbuser: postgres
+  dbpass: postgres
+  dbname: minder
+  sslmode: disable
+
+cursor_file: "/tmp/reminder-cursor"
```

Comment:
@JAORMX suggested storing the cursor state in case the container crashes to which I agreed. Doing so would let the reminder continue where it left from (approx). The benefit is that reminder wouldn't have to start again from the initial entities.

> I'm concerned about state that would make it difficult to run multiple instances in parallel should we need it in the future

My scaling knowledge is a lot inferior to yours. But what I was thinking was to assign each reminder instance a chunk of data from db (somehow), and that was my naive scaling solution here.

Reply:
Thinking out loud about recovery after a crash, there are a few options:

1. We pick a random offset and continue from there.  Tradeoff: some projects might not be scanned for some longer time.
2. We store the offset in a local file (or in the database) so we can resume at the exact right spot.
3. We find some property of the data that allows us to "naturally" pick up at the right spot.

The problem with (3) is that it doesn't really allow for multiple processes.  Option 2 allows for an exact resume if we want, but adds some amount of state management (if we need to keep the file) or devolves into a complicated version of (1), if we allow the file to be blown away.

Scaling:

If we assume a database with 2M entities (we're still a ways off, and other things are gonna break first), then we'd need to scan around 2M / 86400 ~= 24 entities / sec to scan all the entities once a day with a single process.  If our "must revisit" interval is 3 days, we'd end up sending around 600M / day for revisit, assuming that around 10% (200k) are active enough to have had activity in that 3-day period.  Smaller numbers are, of course, easier.  Reading around 25 entities / sec, particularly if it's from a B-Tree index is not really a problem for Postgres, though the follow-on work may be more exciting, particularly if it needs to write a number of rule evaluation results to different parts of the table space.  I'm not up on how well Postgres / AWS Aurora has adapted to the availability of SSD -- discontiguous writes may be mostly a non-issue.

I think we could leave the file if you want, though the code should assume that the file could be empty.